### ISSUE TITLE
Fetching ICY metadata fails on some streams

### ISSUE THREAD

**Issue by nightwolf738** [google/ExoPlayer#5876]
Hello,

I have been waiting for 2.10.0 release for a long time. Finally it has been released. Thanks for your effort.

I tried to fetch ICY metadata from radio streams, but it seems ExoPlayer fails on some streams. Here is some examples:

[http://voyagesc.radyotvonline.com/](http://voyagesc.radyotvonline.com/)
ExoPlayer fails in fetching ICY metadata on this radio stream while 'Current Song' section is not empty when you open link on browser (Chrome in my case). Also, ExoPlayer logs this issue as 

> W/IcyDecoder: Unrecognized ICY tag: null

I would like to also give a working example:
[http://stream.radyoalaturka.com.tr:9100/](http://stream.radyoalaturka.com.tr:9100/)
ExoPlayer successfuly fetches ICY metadata from this stream and invokes `MetadataOutput` listener.

The only difference between two streams is SHOUTcast Server version; failing stream version is 1.9.8 while working stream version is 2.5.5.733

I used an API level 28 Android 9 Emulator
I fetch ICY Metadata as follows:

```
player.addMetadataOutput(metadata -> {
            if ((metadata != null)) {
                final int length = metadata.length();
                if (length > 0) {
                    for (int i = 0; i < length; i++) {
                        final Metadata.Entry entry = metadata.get(i);
                        if (entry instanceof IcyInfo) {
                            final IcyInfo icyInfo = ((IcyInfo) entry);
                            Log.d("IcyInfo", icyInfo.title);
                        } else if (entry instanceof IcyHeaders) {
                            final IcyHeaders icyHeaders = ((IcyHeaders) entry);
                            Log.d("IcyHeaders", icyHeaders.name);
                            Log.d("IcyHeaders", icyHeaders.genre);
                        }
                    }
                }
            }
        });
```

I hope these examples help for finding the issue. If you think I miss something to fetch the metadata, please inform me.

Thank you.

**Comment by ojw28** [comment]
The problem is just that the first stream contains line terminators in the stream title. For example:
```
DJ MERT LEVENT - BON VOYAGE

 - 
```
The `Pattern` we use for matching doesn't handle that. It's easy to fix (note that it'll be working as intended that there are line terminators in the stream title, since that's what's in the underlying stream).

**Comment by egemenhamutcu** [comment]
Thank you for such a short response. I would like to furher this issue to maintain the feature with minimum bugs:

- Fetching ICY metadata does not work at all in my `Sony E5553 Android 6.0, API 23` device, `MetadataOutput` listener never be invoked, while API 26 & API 28 devices works fine. 

- [http://46.20.3.228/](http://46.20.3.228/) this radio stream has an empty 'Current Song' section but ExoPlayer invokes `MetadataOutput` with cyInfo`

**Comment by nightwolf738** [comment]
Thank you for your response in such a short time. I would like to furher this issue to maintain the feature with minimum bugs:

- Fetching ICY metadata does not work at all in my `Sony E5553 Android 6.0, API 23` device, `MetadataOutput` listener is never being invoked, while API 26 & API 28 devices work fine on the same stream. You can test the issue via [http://17773.live.streamtheworld.com/JOY_TURK2AAC_SC](http://17773.live.streamtheworld.com/JOY_TURK2AAC_SC)

- [http://46.20.3.228/](http://46.20.3.228/) this radio stream has an empty 'Current Song' section but ExoPlayer invokes `MetadataOutput` with `IcyInfo` containing title field: `';StreamUrl='`. It seems like ExoPlayer wrongly parses metadata because there is also a 'Stream Url' section when you open the link in browser.

That's all I have found so far. I would like to continue to report more bugs as I face.

Thank you.

**Comment by ojw28** [comment]
> http://46.20.3.228/ this radio stream has an empty 'Current Song' section but ExoPlayer invokes MetadataOutput with IcyInfo containing title field: ';StreamUrl='. It seems like ExoPlayer wrongly parses metadata because there is also a 'Stream Url' section when you open the link in browser.

This is because our `Pattern` doesn't expect empty values, and is also easy to fix.

> Fetching ICY metadata does not work at all in my Sony E5553 Android 6.0, API 23 device, MetadataOutput listener is never being invoked, while API 26 & API 28 devices work fine on the same stream. You can test the issue via http://17773.live.streamtheworld.com/JOY_TURK2AAC_SC

I'm less sure about this one. Does playback of the stream work at all? Does using the OkHttp extension rather than the default network stack resolve the problem? You should probably be using the OkHttp extension anyway if you're playing ICY streams, since on some devices the default network stack will not handle not-strictly-http status responses, and will fail playback with "Unexpected status line: ICY 200 OK" for some ICY streams. We should add an entry to our troubleshooting page about this!

**Comment by nightwolf738** [comment]
Update

Using OkHttp extension resolves the problem in `Sony E5553 Android 6.0, API 23`. I only changed my code from this:

```
public HttpDataSource.Factory buildHttpDataSourceFactory() {
    return new DefaultHttpDataSourceFactory(userAgent);
}
```

to this:

```
public HttpDataSource.Factory buildHttpDataSourceFactory() {
        return new OkHttpDataSourceFactory(new OkHttpClient(), userAgent);
 }
```

We appreciate when other easy-to-fix bugs are resolved.

Thank you.
